// @ton-community/test-utils/dist/test/transaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.compareTransactionForTest = exports.compareTransaction = void 0;
const ton_core_1 = require("ton-core");
const node_inspect_extracted_1 = require("node-inspect-extracted");
function flattenTransaction(tx) {
    return {
        from: tx.inMessage.info.src instanceof ton_core_1.Address ? tx.inMessage.info.src : undefined,
        to: tx.inMessage.info.dest,
        value: tx.inMessage.info.type === 'internal' ? tx.inMessage.info.value.coins : undefined,
        body: tx.inMessage.body,
        initData: tx.inMessage.init?.data ?? undefined,
        initCode: tx.inMessage.init?.code ?? undefined,
        deploy: !!tx.inMessage.init && tx.oldStatus !== 'active' && tx.endStatus === 'active',
        lt: tx.lt,
        now: tx.now,
        outMessagesCount: tx.outMessagesCount,
        oldStatus: tx.oldStatus,
        endStatus: tx.endStatus,
        totalFees: tx.totalFees.coins,
        ...(tx.description.type === 'generic' ? {
            aborted: tx.description.aborted,
            destroyed: tx.description.destroyed,
            exitCode: tx.description.computePhase.type === 'vm' ? tx.description.computePhase.exitCode : undefined,
            success: tx.description.computePhase.type === 'vm'
                ? (tx.description.computePhase.success && tx.description.actionPhase?.success)
                : false,
        } : {
            aborted: undefined,
            destroyed: undefined,
            exitCode: undefined,
            success: undefined,
        }),
    };
}
function compareValue(a, b) {
    if (a instanceof ton_core_1.Address) {
        if (!(b instanceof ton_core_1.Address))
            return false;
        return a.equals(b);
    }
    if (a instanceof ton_core_1.Cell) {
        if (!(b instanceof ton_core_1.Cell))
            return false;
        return a.equals(b);
    }
    return a === b;
}
function compareTransaction(tx, cmp) {
    for (const key in cmp) {
        if (!(key in tx))
            throw new Error(`Unknown flat transaction object key ${key}`);
        const cmpv = cmp[key];
        const txv = tx[key];
        if (typeof cmpv === 'function') {
            if (!cmpv(txv))
                return false;
        }
        else {
            if (!compareValue(txv, cmpv))
                return false;
        }
    }
    return true;
}
exports.compareTransaction = compareTransaction;
function compareTransactionForTest(subject, cmp) {
    if (Array.isArray(subject)) {
        const arr = subject.map(tx => flattenTransaction(tx));
        return {
            pass: arr.some(ftx => compareTransaction(ftx, cmp)),
            posMessage: `Expected ${(0, node_inspect_extracted_1.inspect)(arr)} to contain a transaction that matches pattern ${(0, node_inspect_extracted_1.inspect)(cmp)}`,
            negMessage: `Expected ${(0, node_inspect_extracted_1.inspect)(arr)} NOT to contain a transaction that matches pattern ${(0, node_inspect_extracted_1.inspect)(cmp)}, but it does`,
        };
    }
    else {
        const flat = flattenTransaction(subject);
        return {
            pass: compareTransaction(flat, cmp),
            posMessage: `Expected ${(0, node_inspect_extracted_1.inspect)(flat)} to match pattern ${(0, node_inspect_extracted_1.inspect)(cmp)}`,
            negMessage: `Expected ${(0, node_inspect_extracted_1.inspect)(flat)} NOT to match pattern ${(0, node_inspect_extracted_1.inspect)(cmp)}, but it does`,
        };
    }
}
exports.compareTransactionForTest = compareTransactionForTest;
